src/coord/src/coord.rs

```diff
@@ -393,41 +406,65 @@ impl<C> Coordinator<C>
+            // You will need to read those writes, which happened at the last
+            // known open time. This also means you will need to advance to
+            // those writes, i.e. close over `last_open_local_ts`.
+            self.read_at_offset_from_open_local = 0;
+            self.writes_at_open_ts = false;
+        }
+
+        self.last_open_local_ts - self.read_at_offset_from_open_local
     }
 
-    /// Assign a timestamp for a write. Writes following reads must ensure that they are assigned a
-    /// strictly larger timestamp to ensure they are not visible to any real-time earlier reads.
-    fn get_write_ts(&mut self) -> Timestamp {
-        let ts = if self.last_op_was_read {
-            self.last_op_was_read = false;
-            cmp::max(self.get_ts(), self.read_lower_bound + 1)
-        } else {
-            self.get_ts()
-        };
-        self.read_lower_bound = cmp::max(ts, self.closed_up_to);
-        self.read_lower_bound
+    /// Assign a timestamp for a write to a local input. Writes following reads
+    /// must ensure that they are assigned a strictly larger timestamp to ensure
+    /// they are not visible to any real-time earlier reads.
+    fn get_local_write_ts(&mut self) -> Timestamp {
+        // This assert is valid because:
+        // - Whenever a write precedes a read, the read sets
+        //   `read_at_offset_from_open_local = 0`, which will advance the
+        //   `last_open_local_ts`.
+        // - Reads always enter a greenthread, letting the Coordinator see that
+        //   the `last_open_local_ts` needs to advance before having the
+        //   opportunity to generate any write timestamps.
+        // - Advancing the timestamp sets `read_at_offset_from_open_local = 1`.
+        assert!(
+            self.read_at_offset_from_open_local > 0,
+            "do not perform writes at time where tables want to read"
+        );
+
+        self.writes_at_open_ts = true;
+
+        self.last_open_local_ts
     }
 
-    /// Fetch a new timestamp.
-    fn get_ts(&mut self) -> Timestamp {
-        // Next time we have a chance, we will force all local inputs forward.
-        self.need_advance = true;
+    /// Opens a new timestamp for local inputs at which writes may occur, and
+    /// where reads should return quickly at a value 1 less.
+    fn open_new_local_ts(&mut self) {
         // This is a hack. In a perfect world we would represent time as having a "real" dimension
         // and a "coordinator" dimension so that clients always observed linearizability from
         // things the coordinator did without being related to the real dimension.
         let ts = (self.catalog.config().now)();
 
-        if ts < self.read_lower_bound {
-            self.read_lower_bound
-        } else {
-            ts
-        }
+        // We cannot depend on `self.catalog.config().now`'s value to change
+        // when invoked (this feature enables us to drive the Coordinator's time
```

Comment:
We also need this incase the system time has gone backward for any reason in the non-coordtest case.

Reply:
Poorly worded; fixed.